csharp/src/Ice/IPEndpoint.cs

```diff
@@ -101,41 +102,93 @@ public override bool IsLocal(Endpoint endpoint)
-                                                                     networkProxy?.IPVersion ?? Network.EnableBoth,
-                                                                     cancel).ConfigureAwait(false);
-                return addrs.Select(item => CreateConnector(item, networkProxy));
+                addresses = await Network.GetAddressesForClientEndpointAsync(
+                    Host,
+                    Port,
+                    networkProxy?.IPVersion ?? Network.EnableBoth,
+                    cancel).ConfigureAwait(false);
             }
             catch (Exception ex)
             {
-                observer?.Failed(ex.GetType().FullName ?? "System.Exception");
+                endpointLookupObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
                 throw;
             }
             finally
             {
-                observer?.Detach();
+                endpointLookupObserver?.Detach();
+            }
+
+            IPEndPoint lastAddress = addresses[^1];
+            Connection? connection = null;
+            foreach (IPEndPoint address in addresses)
+            {
+                IObserver? connectionEstablishmentObserver =
+                    Communicator.Observer?.GetConnectionEstablishmentObserver(this, address.ToString());
+                connectionEstablishmentObserver?.Attach();
+                try
+                {
+                    bool secureOnly = preferNonSecure switch
+                    {
+                        NonSecure.SameHost => !address.IsSameHost(),
+                        NonSecure.TrustedHost => true, // TODO check if address is a trusted host
+                        NonSecure.Always => false,
+                        _ => true
+                    };
+                    connection = CreateConnection(preferNonSecure, secureOnly, address, networkProxy, cookie);
+                    await connection.InitializeAsync().ConfigureAwait(false);
+                }
+                catch (OperationCanceledException ex)
+                {
+                    connectionEstablishmentObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
+                    throw;
+                }
+                catch (Exception ex)
+                {
+                    connectionEstablishmentObserver?.Failed(ex.GetType().FullName ?? "System.Exception");
+                    // Ignore the exception unless this is the last address
+                    if (ReferenceEquals(lastAddress, address))
+                    {
+                        throw;
+                    }
+                }
+                finally
+                {
+                    connectionEstablishmentObserver?.Detach();
+                }
             }
+            Debug.Assert(connection != null);
+            return connection;
+        }
+
+        protected internal abstract Connection CreateConnection(
+            NonSecure preferNonSecure,
```

Comment:
I find this preferNonSecure + secureOnly parameters confusing. Why do we need both?

Reply:
We have to keep the `preferNonSecure` to be able to remove the connection from the map

Reply:
For me `secureOnly == (preferNonSecure == NonSecure.Never)`. What's the difference here?

Reply:
It can also mean that (preferNonSecure == NonSecure.SameHost) and the endpoint doesn't represent the same host